src/language-js/needs-parens.js

```diff
@@ -1258,4 +1228,24 @@ function shouldAddParenthesesToChainElement(path) {
   return false;
 }
 
+function canDecoratorExpressionUnparenthesized(node) {
+  if (node.type === "ChainExpression") {
+    node = node.expression;
+  }
+
+  if (node.type === "Identifier") {
+    return true;
+  }
+
+  if (isCallExpression(node)) {
+    return node.callee.type === "Identifier";
+  }
+
+  if (isMemberExpression(node)) {
+    return !node.computed && node.object.type === "Identifier";
+  }
+
+  return false;
```

Comment:
@y-schneider It seems that only these cases don't need parentheses?

Reply:
That is my understanding as well. Thank you so much!

Reply:
The logic has changed a little bit since I posted the comment.

I saw babel has this logic https://github.com/babel/babel/pull/16627 😄  They understand the AST better.

Reply:
Oh, I didn't realise the PR shows the original code. I went straight to the code to see it in context and tried it in the playground as well 👍